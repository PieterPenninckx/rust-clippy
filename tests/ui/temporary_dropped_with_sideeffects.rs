// #![cfg_attr(feature = "clippy", deny(dropping_temporary_with_side_effect))]
#![allow(mutex_atomic)]
#![feature(tool_attributes)]
#![feature(stmt_expr_attributes)]

#![deny(dropping_temporary_with_side_effect)]

#[allow(unused_imports)]
#[allow(unused_variables)]
use std::sync::{Mutex, MutexGuard};

use std::ops::Index;
use std::sync::RwLock;
use std::cell::RefCell;

trait MyTrait {
	type AssociatedType;
	fn get(self) -> Self::AssociatedType;
}

#[derive(Clone, Copy, Debug)]
struct S {
	x: u8
}

struct MutexContainer<'a> {
	rwlmg: MutexGuard<'a, S>
}

impl<'a> MyTrait for MutexContainer<'a> {
	type AssociatedType = MutexGuard<'a, S>;
	fn get(self) -> MutexGuard<'a, S> {
		self.rwlmg // Should not trigger
	}
}

impl<'a> MutexContainer<'a> {
	#[allow(dead_code)]
	fn get_wrapped(self) -> MutexGuard<'a, S> {
		self.get() // Should not trigger
	}
}

#[allow(needless_lifetimes)]
#[allow(dead_code)]
fn get<'a>(m: MutexContainer<'a>) -> MutexGuard<'a, S> {
	m.get() // Should not trigger
}

#[allow(dead_code)]
fn g<T: MyTrait> (param: T) {
	let _ = param.get(); // Should not trigger.
}

#[allow(dead_code)]
fn f<'a, T: MyTrait<AssociatedType = MutexGuard<'a, S>>> (param: T) {
	let _ = param.get();  // Should trigger.
}

struct MutexIndexable {
}

impl<'a, T> Index<&'a MutexGuard<'a, T>> for MutexIndexable {
	type Output = u8;
	fn index(&self, _mg: &'a MutexGuard<T>) -> &u8 { &5 }
}

fn should_trigger() {
	let mutex = Mutex::new(S{x: 4});
	
	let _   = mutex.lock();     // Should trigger.
	let _   = (mutex.lock(), ); // Should trigger.
	let _   = [mutex.lock()];   // Should trigger.
	let _   = {mutex.lock()};   // Should trigger.
	let _a1 = mutex.lock().unwrap().x; // Should trigger.
	let _a2 = MutexContainer{rwlmg: mutex.lock().unwrap()}; // Should trigger
	let _a3 = [1,2,3,4,5][mutex.lock().unwrap().x as usize]; // Should trigger
	let _a4 = MutexIndexable{}[&mutex.lock().unwrap()]; // Should trigger
	let _a5 = *mutex.lock().unwrap(); // Should trigger
	let _a6 = [mutex.lock(), mutex.lock()].len(); // Should trigger
	println!("{:?}", (mutex.lock(), 4)); // Should trigger
	let _a7 = (&mutex.lock()).is_err(); // Should trigger
	
	let mutex_bool = Mutex::new(true);
	if *mutex_bool.lock().unwrap() { // Should trigger
		println!("Do something.");
	}

	match mutex_bool.lock() { // should trigger
		Ok(ref x) if **x => {
			println!("Do something.");
		}
		_ => ()
	}
	
	let rwlock = RwLock::new(0);
	let _ = rwlock.read();  // Should trigger
	let _ = rwlock.write(); // Should trigger
	
	let refcell = RefCell::new(0);
	let _ = refcell.borrow();     // Should trigger
	let _ = refcell.borrow_mut(); // Should trigger
	
	
	let cache: RwLock<Option<u8>> = RwLock::new(None);
	let _value = {
		match cache.read() { // Should trigger
			Ok(ref y) if y.is_some() => Some(y.unwrap()),
			_ => None,
		}
	}
	.unwrap_or_else(|| {
		let mut _data = cache.write().unwrap();
		*_data = Some(0);
		0
    });
}

fn should_not_trigger() {
	let mutex = Mutex::new(S{x: 4});
	let _a1 = mutex.lock().unwrap(); // Should not trigger
	
	let mutex_bool = Mutex::new(true);
	let must_do_something : bool;
	{
		let value = mutex_bool.lock().unwrap(); // Should not trigger
		must_do_something = *value;
	}
	if must_do_something {
		println!("Do something.");
	}
	
	{
		let value = mutex_bool.lock();
		match value {
			Ok(ref x) if **x => {
					println!("Do something");
			},
			_ => ()
		}
	}
	
	{
		let _data = mutex.lock().unwrap();
		println!("Do something with data.");
	}
}

fn main() {
	should_trigger();
	should_not_trigger();
}
