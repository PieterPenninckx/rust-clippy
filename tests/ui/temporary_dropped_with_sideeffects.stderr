error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:57:10
   |
57 |     let _ = param.get();  // Should trigger.
   |             ^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/temporary_dropped_with_sideeffects.rs:6:9
   |
6  | #![deny(dropping_temporary_with_side_effect)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:71:12
   |
71 |     let _   = mutex.lock();     // Should trigger.
   |               ^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:72:13
   |
72 |     let _   = (mutex.lock(), ); // Should trigger.
   |                ^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:73:13
   |
73 |     let _   = [mutex.lock()];   // Should trigger.
   |                ^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:74:12
   |
74 |     let _   = {mutex.lock()};   // Should trigger.
   |               ^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:74:13
   |
74 |     let _   = {mutex.lock()};   // Should trigger.
   |                ^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:75:12
   |
75 |     let _a1 = mutex.lock().unwrap().x; // Should trigger.
   |               ^^^^^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:76:34
   |
76 |     let _a2 = MutexContainer{rwlmg: mutex.lock().unwrap()}; // Should trigger
   |                                     ^^^^^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:77:24
   |
77 |     let _a3 = [1,2,3,4,5][mutex.lock().unwrap().x as usize]; // Should trigger
   |                           ^^^^^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:78:30
   |
78 |     let _a4 = MutexIndexable{}[&mutex.lock().unwrap()]; // Should trigger
   |                                 ^^^^^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:79:13
   |
79 |     let _a5 = *mutex.lock().unwrap(); // Should trigger
   |                ^^^^^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:80:13
   |
80 |     let _a6 = [mutex.lock(), mutex.lock()].len(); // Should trigger
   |                ^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:80:27
   |
80 |     let _a6 = [mutex.lock(), mutex.lock()].len(); // Should trigger
   |                              ^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:81:20
   |
81 |     println!("{:?}", (mutex.lock(), 4)); // Should trigger
   |                       ^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:82:14
   |
82 |     let _a7 = (&mutex.lock()).is_err(); // Should trigger
   |                 ^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:85:6
   |
85 |     if *mutex_bool.lock().unwrap() { // Should trigger
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:89:8
   |
89 |     match mutex_bool.lock() { // should trigger
   |           ^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:97:10
   |
97 |     let _ = rwlock.read();  // Should trigger
   |             ^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
  --> $DIR/temporary_dropped_with_sideeffects.rs:98:10
   |
98 |     let _ = rwlock.write(); // Should trigger
   |             ^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
   --> $DIR/temporary_dropped_with_sideeffects.rs:101:10
    |
101 |     let _ = refcell.borrow();     // Should trigger
    |             ^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
   --> $DIR/temporary_dropped_with_sideeffects.rs:102:10
    |
102 |     let _ = refcell.borrow_mut(); // Should trigger
    |             ^^^^^^^^^^^^^^^^^^^^

error: Expression results in an anonymous temporary variable that will have side effects when being dropped.
   --> $DIR/temporary_dropped_with_sideeffects.rs:107:9
    |
107 |         match cache.read() { // Should trigger
    |               ^^^^^^^^^^^^

error: aborting due to 22 previous errors

